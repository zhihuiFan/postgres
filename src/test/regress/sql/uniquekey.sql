CREATE TABLE uk_t (id int primary key, a int not null, b int not null, c int, d int, e int);

explain (costs off)
select distinct id from uk_t;

explain (costs off)
select distinct e from uk_t where id = e;

create unique index on uk_t (a, b);
create unique index on uk_t (c, d);

explain (costs off)
select distinct a, b from uk_t;

explain (costs off)
select distinct c, d from uk_t;

explain (costs off)
select distinct c, d from uk_t
where c > 0 and d > 0;

explain (costs off)
select distinct d from uk_t
where c > 1 and d > 0;

explain (costs off)
select distinct d from uk_t
where c = 1 and d > 0;





