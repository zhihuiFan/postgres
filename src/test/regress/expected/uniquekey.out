CREATE TABLE uqk1(a int, pk int primary key, c int,  d int);
CREATE TABLE uqk2(a int, pk int primary key, c int,  d int);
INSERT INTO uqk1 VALUES(1, 1, 1, 1), (2, 2, 2, 2), (3, 3, 3, 3), (4, 4, null, 4), (5, 5, null, 4);
INSERT INTO uqk2 VALUES(1, 1, 1, 1), (4, 4, 4, 4), (5, 5, 5, 5);
ANALYZE uqk1;
ANALYZE uqk2;
-- Test single table primary key.
EXPLAIN (COSTS OFF) SELECT DISTINCT * FROM uqk1;
    QUERY PLAN    
------------------
 Seq Scan on uqk1
(1 row)

-- Test EC case.
EXPLAIN (COSTS OFF) SELECT DISTINCT d FROM uqk1 WHERE d = pk;
     QUERY PLAN     
--------------------
 Seq Scan on uqk1
   Filter: (d = pk)
(2 rows)

-- Test UniqueKey indexes.
CREATE UNIQUE INDEX uqk1_ukcd ON uqk1(c, d);
-- Test not null quals and not null per catalog.
EXPLAIN (COSTS OFF) SELECT DISTINCT c, d FROM uqk1;
       QUERY PLAN       
------------------------
 HashAggregate
   Group Key: c, d
   ->  Seq Scan on uqk1
(3 rows)

EXPLAIN (COSTS OFF) SELECT DISTINCT c, d FROM uqk1 WHERE c is NOT NULL;
              QUERY PLAN               
---------------------------------------
 Unique
   ->  Sort
         Sort Key: c, d
         ->  Seq Scan on uqk1
               Filter: (c IS NOT NULL)
(5 rows)

EXPLAIN (COSTS OFF) SELECT DISTINCT d FROM uqk1 WHERE c = 1;
          QUERY PLAN           
-------------------------------
 Unique
   ->  Sort
         Sort Key: d
         ->  Seq Scan on uqk1
               Filter: (c = 1)
(5 rows)

ALTER TABLE uqk1 ALTER COLUMN d SET NOT NULL;
EXPLAIN (COSTS OFF) SELECT DISTINCT c, d FROM uqk1 WHERE c is NOT NULL;
        QUERY PLAN         
---------------------------
 Seq Scan on uqk1
   Filter: (c IS NOT NULL)
(2 rows)

-- Test UniqueKey column reduction.
EXPLAIN (COSTS OFF) SELECT DISTINCT d FROM uqk1 WHERE c = 1;
    QUERY PLAN     
-------------------
 Seq Scan on uqk1
   Filter: (c = 1)
(2 rows)

EXPLAIN (COSTS OFF) SELECT DISTINCT a FROM uqk1 WHERE c = 1 and d = 1;
           QUERY PLAN            
---------------------------------
 Seq Scan on uqk1
   Filter: ((c = 1) AND (d = 1))
(2 rows)

-- Test Distinct ON
EXPLAIN (COSTS OFF) SELECT DISTINCT ON(pk) d FROM uqk1;
    QUERY PLAN    
------------------
 Seq Scan on uqk1
(1 row)

